discord/object.py

```diff
@@ -77,19 +79,35 @@ class Object(Hashable):
     -----------
     id: :class:`int`
         The ID of the object.
+    type: Type[:class:`abc.Snowflake`]
+        The discord.py model type of the object, if not specified, defaults to this class.
+
+        .. note::
+
+            In instances where there are multiple applicable types, use a shared base class.
+            for example, both :class:`Member` and :class:`User` are subclasses of :class:`abc.User`.
+
+        .. versionadded:: 2.0
     """
 
-    def __init__(self, id: SupportsIntCast):
+    def __init__(self, id: SupportsIntCast, *, type: Type[abc.Snowflake] = MISSING):
         try:
             id = int(id)
         except ValueError:
             raise TypeError(f'id parameter must be convertable to int not {id.__class__!r}') from None
-        else:
-            self.id = id
+        self.id: int = id
+        self.type: Type[abc.Snowflake] = type or self.__class__
```

Comment:
Similarly to `id` I would like this to be validated, however without moving `abc.Snowflake` into here, or its own file, it can't be done as `abc` calls `Object.__init__`.